Searching for publication year ranges broken

Searching for publication year like `>2016` is broken. Results in a NumberFormatException

```
Caused by: java.lang.NumberFormatException: For input string: ">"
        at java.base/java.lang.NumberFormatException.forInputString(NumberFormatException.java:65) ~[na:na]
        at java.base/java.lang.Integer.parseInt(Integer.java:652) ~[na:na]
        at java.base/java.lang.Integer.parseInt(Integer.java:770) ~[na:na]
        at ch.difty.scipamato.core.entity.search.IntegerSearchTerm.<init>(IntegerSearchTerm.java:108) ~[scipamato-core-entity-1.1.6.jar!/:1.1.6]
        at ch.difty.scipamato.core.entity.search.IntegerSearchTerm.<init>(IntegerSearchTerm.java:54) ~[scipamato-core-entity-1.1.6.jar!/:1.1.6]
        at ch.difty.scipamato.core.entity.search.IntegerSearchTerm.<init>(IntegerSearchTerm.java:50) ~[scipamato-core-entity-1.1.6.jar!/:1.1.6]
        at ch.difty.scipamato.core.entity.search.SearchTerm.newIntegerSearchTerm(SearchTerm.java:85) ~[scipamato-core-entity-1.1.6.jar!/:1.1.6]
        at ch.difty.scipamato.core.entity.search.SearchCondition.setIntegerValue(SearchCondition.java:475) ~[scipamato-core-entity-1.1.6.jar!/:1.1.6]
        at ch.difty.scipamato.core.entity.search.SearchCondition.setPublicationYear(SearchCondition.java:216) ~[scipamato-core-entity-1.1.6.jar!/:1.1.6]
        at jdk.internal.reflect.GeneratedMethodAccessor516.invoke(Unknown Source) ~[na:na]
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[na:na]
        at java.base/java.lang.reflect.Method.invoke(Method.java:566) ~[na:na]
        at org.apache.wicket.core.util.lang.PropertyResolver$MethodGetAndSet.setValue(PropertyResolver.java:999) ~[wicket-core-8.3.0.jar!/:8.3.0]
        ... 116 common frames omitted
```

--- a/CHANGELOG.asciidoc
+++ b/CHANGELOG.asciidoc
@@ -72,6 +72,7 @@ NOTE: References to user stories are in the form Iteration/Story-Number.
 .Removed
 
 .Fixed
+- {url-issues}82[#82] - Fix searching for publication year ranges and for ID
 
 .Security
 

--- a/implementation/scipamato/scipamato-core-entity/src/main/java/ch/difty/scipamato/core/entity/search/IntegerSearchTerm.java
+++ b/implementation/scipamato/scipamato-core-entity/src/main/java/ch/difty/scipamato/core/entity/search/IntegerSearchTerm.java
@@ -1,5 +1,9 @@
 package ch.difty.scipamato.core.entity.search;
 
+import java.util.Arrays;
+
+import org.apache.commons.lang3.math.NumberUtils;
+
 /**
  * Implementation of {@link AbstractSearchTerm} working with Integer fields. The
  * following {@link MatchType}s are implemented:
@@ -13,6 +17,7 @@ package ch.difty.scipamato.core.entity.search;
  * <li><b>RANGE:</b> {@code  2014-2017 }</li>
  * <li><b>MISSING:</b> the field has no value.</li>
  * <li><b>PRESENT:</b> the field has any value.</li>
+ * <li><b>INCOMPLETE:</b> the search term ist not complete yet (partial entry).</li>
  * </ul>
  * <p>
  * All rawValues and their individual parts are trimmed, so the following
@@ -23,6 +28,9 @@ package ch.difty.scipamato.core.entity.search;
  * <li>{@code <= 2014 }</li>
  * <li>{@code  2014  - 2017 }</li>
  * </ul>
+ * <p>
+ * NOTE: The INCOMPLETE will not find any papers. It's more about gracefully handling
+ * unfinished search terms. See https://github.com/ursjoss/scipamato/issues/84
  *
  * @author u.joss
  */
@@ -39,7 +47,23 @@ public class IntegerSearchTerm extends AbstractSearchTerm {
         LESS_OR_EQUAL,
         RANGE,
         MISSING,
-        PRESENT
+        PRESENT,
+        INCOMPLETE
+    }
+
+    private enum Ops {
+        LESS_THAN("<"),
+        LESS_THAN_OR_EQUAL("<="),
+        EQUAL("="),
+        GREATER_THAN_OR_EQUAL(">="),
+        GREATER_THAN(">"),
+        RANGE("-");
+
+        private final String symbol;
+
+        Ops(final String symbol) {
+            this.symbol = symbol;
+        }
     }
 
     private final MatchType type;
@@ -57,7 +81,13 @@ public class IntegerSearchTerm extends AbstractSearchTerm {
     IntegerSearchTerm(final Long id, final Long searchConditionId, final String fieldName, final String rawSearchTerm) {
         super(id, SearchTermType.INTEGER, searchConditionId, fieldName, rawSearchTerm);
         final String rst = rawSearchTerm.trim();
-        if ("=\"\"".equals(rst) || "\"\"".equals(rst)) {
+        if (rst.startsWith(Ops.RANGE.symbol) || rst.endsWith(Ops.RANGE.symbol) || Arrays
+            .stream(Ops.values())
+            .anyMatch(op -> op.symbol.equals(rst))) {
+            this.type = MatchType.INCOMPLETE;
+            this.value = 0;
+            this.value2 = 0;
+        } else if ("=\"\"".equals(rst) || "\"\"".equals(rst)) {
             this.type = MatchType.MISSING;
             this.value = 0;
             this.value2 = 0;
@@ -65,48 +95,54 @@ public class IntegerSearchTerm extends AbstractSearchTerm {
             this.type = MatchType.PRESENT;
             this.value = 0;
             this.value2 = 0;
-        } else if (rst.length() > COMP_SYMBOL_MAX_LENGTH && rst.startsWith(">=")) {
+        } else if (rst.startsWith(Ops.GREATER_THAN_OR_EQUAL.symbol)) {
             this.type = MatchType.GREATER_OR_EQUAL;
             this.value = Integer.parseInt(rst
                 .substring(COMP_SYMBOL_MAX_LENGTH)
                 .trim());
             this.value2 = this.value;
-        } else if (rst.length() > 1 && rst.startsWith(">")) {
+        } else if (rst.startsWith(Ops.GREATER_THAN.symbol)) {
             this.type = MatchType.GREATER_THAN;
             this.value = Integer.parseInt(rst
                 .substring(1)
                 .trim());
             this.value2 = this.value;
-        } else if (rst.length() > COMP_SYMBOL_MAX_LENGTH && rst.startsWith("<=")) {
+        } else if (rst.startsWith(Ops.LESS_THAN_OR_EQUAL.symbol)) {
             this.type = MatchType.LESS_OR_EQUAL;
             this.value = Integer.parseInt(rst
                 .substring(COMP_SYMBOL_MAX_LENGTH)
                 .trim());
             this.value2 = this.value;
-        } else if (rst.length() > 1 && rst.startsWith("<")) {
+        } else if (rst.startsWith(Ops.LESS_THAN.symbol)) {
             this.type = MatchType.LESS_THAN;
             this.value = Integer.parseInt(rst
                 .substring(1)
                 .trim());
             this.value2 = this.value;
-        } else if (rst.length() > 1 && rst.startsWith("=")) {
+        } else if (rst.startsWith(Ops.EQUAL.symbol)) {
             this.type = MatchType.EXACT;
             this.value = Integer.parseInt(rst
                 .substring(1)
                 .trim());
             this.value2 = this.value;
-        } else if (rst.length() > 1 && rst.contains("-")) {
+        } else if (rst.contains(Ops.RANGE.symbol)) {
+            final String symbol = Ops.RANGE.symbol;
+            final int indexOfSymbol = rst.indexOf(symbol);
             this.type = MatchType.RANGE;
             this.value = Integer.parseInt(rst
-                .substring(0, rst.indexOf('-'))
+                .substring(0, indexOfSymbol)
                 .trim());
             this.value2 = Integer.parseInt(rst
-                .substring(rst.indexOf('-') + 1)
+                .substring(indexOfSymbol + 1)
                 .trim());
-        } else {
-            this.type = MatchType.EXACT;
+        } else if (NumberUtils.isCreatable(rst.trim())) {
             this.value = Integer.parseInt(rst.trim());
             this.value2 = this.value;
+            this.type = MatchType.EXACT;
+        } else {
+            this.type = MatchType.INCOMPLETE;
+            this.value = 1; // arbitrary value to make it different from the first block and mute the intellij warning)
+            this.value2 = 1;
         }
     }
 

--- a/implementation/scipamato/scipamato-core-entity/src/test/java/ch/difty/scipamato/core/entity/search/IntegerSearchTermTest.java
+++ b/implementation/scipamato/scipamato-core-entity/src/test/java/ch/difty/scipamato/core/entity/search/IntegerSearchTermTest.java
@@ -129,15 +129,66 @@ class IntegerSearchTermTest {
     }
 
     @Test
+    void incompleteSearchTermEqual() {
+        final String raw = "=";
+        st = new IntegerSearchTerm(CONDITION_ID, FIELD_NAME, raw);
+        assertTerm(MatchType.INCOMPLETE, 0, 0, raw);
+    }
+
+    @Test
+    void incompleteSearchTermRange() {
+        final String raw = "-";
+        st = new IntegerSearchTerm(CONDITION_ID, FIELD_NAME, raw);
+        assertTerm(MatchType.INCOMPLETE, 0, 0, raw);
+    }
+
+    @Test
+    void incompleteSearchTermRangeRighOpen() {
+        final String raw = "2017-";
+        st = new IntegerSearchTerm(CONDITION_ID, FIELD_NAME, raw);
+        assertTerm(MatchType.INCOMPLETE, 0, 0, raw);
+    }
+
+    @Test
+    void incompleteSearchTermRangeLeftOpen() {
+        final String raw = "-2019";
+        st = new IntegerSearchTerm(CONDITION_ID, FIELD_NAME, raw);
+        assertTerm(MatchType.INCOMPLETE, 0, 0, raw);
+    }
+
+    @Test
+    void incompleteSearchTermGreaterThan() {
+        final String raw = ">";
+        st = new IntegerSearchTerm(CONDITION_ID, FIELD_NAME, raw);
+        assertTerm(MatchType.INCOMPLETE, 0, 0, raw);
+    }
+
+    @Test
+    void incompleteSearchTermGreaterThanOrEqual() {
+        final String raw = ">=";
+        st = new IntegerSearchTerm(CONDITION_ID, FIELD_NAME, raw);
+        assertTerm(MatchType.INCOMPLETE, 0, 0, raw);
+    }
+
+    @Test
+    void incompleteSearchTermLessThan() {
+        final String raw = "<";
+        st = new IntegerSearchTerm(CONDITION_ID, FIELD_NAME, raw);
+        assertTerm(MatchType.INCOMPLETE, 0, 0, raw);
+    }
+
+    @Test
+    void incompleteSearchTermLessThanOrEqual() {
+        final String raw = "<=";
+        st = new IntegerSearchTerm(CONDITION_ID, FIELD_NAME, raw);
+        assertTerm(MatchType.INCOMPLETE, 0, 0, raw);
+    }
+
+    @Test
     void invalidSearch_withNonNumericCharacters() {
-        try {
-            new IntegerSearchTerm(CONDITION_ID, FIELD_NAME, "2014a");
-            fail("Should have thrown exception");
-        } catch (Exception ex) {
-            assertThat(ex)
-                .isInstanceOf(NumberFormatException.class)
-                .hasMessage("For input string: \"2014a\"");
-        }
+        final String raw = "2014a";
+        st = new IntegerSearchTerm(CONDITION_ID, FIELD_NAME, raw);
+        assertTerm(MatchType.INCOMPLETE, 1, 1, raw);
     }
 
     @Test

--- a/implementation/scipamato/scipamato-core-persistence-jooq/src/main/java/ch/difty/scipamato/core/persistence/paper/searchorder/IntegerSearchTermEvaluator.java
+++ b/implementation/scipamato/scipamato-core-persistence-jooq/src/main/java/ch/difty/scipamato/core/persistence/paper/searchorder/IntegerSearchTermEvaluator.java
@@ -39,6 +39,8 @@ public class IntegerSearchTermEvaluator implements SearchTermEvaluator<IntegerSe
             return field.isNull();
         case PRESENT:
             return field.isNotNull();
+        case INCOMPLETE:
+            return DSL.falseCondition();
         default:
             return field.equal(value);
         }

--- a/implementation/scipamato/scipamato-core-persistence-jooq/src/main/java/ch/difty/scipamato/core/persistence/paper/searchorder/IntegerSearchTermEvaluator.java
+++ b/implementation/scipamato/scipamato-core-persistence-jooq/src/main/java/ch/difty/scipamato/core/persistence/paper/searchorder/IntegerSearchTermEvaluator.java
@@ -20,7 +20,7 @@ public class IntegerSearchTermEvaluator implements SearchTermEvaluator<IntegerSe
     public Condition evaluate(final IntegerSearchTerm searchTerm) {
         AssertAs.notNull(searchTerm, "searchTerm");
 
-        final String fieldName = TranslationUtils.deCamelCase(searchTerm.getFieldName());
+        final String fieldName = getFieldName(searchTerm.getFieldName());
         final Field<Object> field = DSL.field(fieldName);
         final Param<Integer> value = DSL.val(searchTerm.getValue());
 
@@ -44,6 +44,13 @@ public class IntegerSearchTermEvaluator implements SearchTermEvaluator<IntegerSe
         default:
             return field.equal(value);
         }
+    }
+
+    private String getFieldName(final String fieldName) {
+        if ("id".equals(fieldName))
+            return "paper.id";
+        else
+            return TranslationUtils.deCamelCase(fieldName);
     }
 
 }

--- a/implementation/scipamato/scipamato-core-persistence-jooq/src/test/java/ch/difty/scipamato/core/persistence/paper/searchorder/IntegerSearchTermEvaluatorTest.java
+++ b/implementation/scipamato/scipamato-core-persistence-jooq/src/test/java/ch/difty/scipamato/core/persistence/paper/searchorder/IntegerSearchTermEvaluatorTest.java
@@ -103,4 +103,12 @@ class IntegerSearchTermEvaluatorTest extends SearchTermEvaluatorTest<IntegerSear
         assertThat(c.toString()).isEqualTo("1 = 0");
     }
 
+    @Test
+    void buildingCondition_withFieldId_prependsTable() {
+        expectSearchTerm(MatchType.GREATER_THAN, 10);
+        when(stMock.getFieldName()).thenReturn("id");
+        Condition c = e.evaluate(stMock);
+        assertThat(c.toString()).isEqualTo("paper.id > 10");
+    }
+
 }

--- a/implementation/scipamato/scipamato-core-persistence-jooq/src/test/java/ch/difty/scipamato/core/persistence/paper/searchorder/IntegerSearchTermEvaluatorTest.java
+++ b/implementation/scipamato/scipamato-core-persistence-jooq/src/test/java/ch/difty/scipamato/core/persistence/paper/searchorder/IntegerSearchTermEvaluatorTest.java
@@ -96,4 +96,11 @@ class IntegerSearchTermEvaluatorTest extends SearchTermEvaluatorTest<IntegerSear
         assertThat(c.toString()).isEqualTo("field_x is not null");
     }
 
+    @Test
+    void buildingCondition_withIncompleteSearchTerm() {
+        when(stMock.getType()).thenReturn(MatchType.INCOMPLETE);
+        Condition c = e.evaluate(stMock);
+        assertThat(c.toString()).isEqualTo("1 = 0");
+    }
+
 }

